File: LeetCode/199/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  vector<int> rightSideView(TreeNode *root) {
    if (root == nullptr)
      return {};

    vector<int> ans;
    queue<TreeNode *> q;
    q.push(root);

    // Level order traversal
    while (!q.empty()) {

      int sz = q.size();
      int visible = 0;
      for (int i = 0; i < sz; ++i) {

        TreeNode *node = q.front();
        q.pop();
        visible = node->val;

        if (node->left)
          q.push(node->left);
        if (node->right)
          q.push(node->right);
      }

      // At each level, only 1 will be visible and that's the last
      // one added
      ans.emplace_back(visible);
    }
    return ans;
  }
};

File: LeetCode/199/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rightSideView(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	ans := []int{}
	q := []*TreeNode{}
	q = append(q, root)

	for len(q) > 0 {
		sz, visible := len(q), 0
		for i := 0; i < sz; i++ {
			ele := q[0]
			visible = ele.Val
			q = q[1:]
			if ele.Left != nil {
				q = append(q, ele.Left)
			}
			if ele.Right != nil {
				q = append(q, ele.Right)
			}
		}
		ans = append(ans, visible)
	}
	return ans

}